src/lib/middleware/rbac-middleware.ts

```diff
@@ -0,0 +1,53 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+import {
+    CREATE_FEATURE,
+    UPDATE_FEATURE,
+    DELETE_FEATURE,
+    ADMIN,
+} from '../permissions';
+
+import { isRbacEnabled } from '../util/feature-enabled';
+
+const rbacMiddleware = (config: any, { accessService }: any): any => {
+    if (!isRbacEnabled(config)) {
+        return (req, res, next) => next();
+    }
+
+    const logger = config.getLogger('/middleware/rbac-middleware.js');
+    logger.info('Enabling RBAC');
+
+    const { featureToggleStore } = config.stores;
+
+    return (req, res, next) => {
+        req.checkRbac = async (permission: string) => {
+            const { user, params } = req;
+
+            // Support ADMIN API tokens for enterpriseAuthentication.
+            if (user && user.isAPI) {
+                return user.permissions.includes(ADMIN);
+            }
+
+            if (!user || !user.id) {
+                logger.error('RBAC requires users to have a unique id.');
+                return false;
+            }
+
+            // For /api/admin/projects/:projectId we will find it as part of params
+            let { projectId } = params;
+
+            // Temporary workaround to figure our projectId for feature toggle updates.
+            if ([UPDATE_FEATURE, DELETE_FEATURE].includes(permission)) {
+                const { featureName } = params;
+                projectId = await featureToggleStore.getProjectId(featureName);
+            } else if (permission === CREATE_FEATURE) {
+                projectId = req.body.project;
+            }
+
+            return accessService.hasPermission(user, permission, projectId);
+        };
```

Comment:
I need a double-take when I read the `if (!user || !user.id)` check. Seeing as how you've already checked for user on L26, may I suggest something like this?
```suggestion
        req.checkRbac = async (permission: string) => {
            const { user, params } = req;

            // Support ADMIN API tokens for enterpriseAuthentication.
            if (user) {
                if(user.isAPI) {
                 return user.permissions.includes(ADMIN);
                }
                if (!user.id) {
                  logger.error('RBAC requires users to have a unique id.');
                }
               // For /api/admin/projects/:projectId we will find it as part of params
               let { projectId } = params;

               // Temporary workaround to figure our projectId for feature toggle updates.
               if ([UPDATE_FEATURE, DELETE_FEATURE].includes(permission)) {
                   const { featureName } = params;
                   projectId = await featureToggleStore.getProjectId(featureName);
               } else if (permission === CREATE_FEATURE) {
                  projectId = req.body.project;
               }
               return accessService.hasPermission(user, permission, projectId);    
            } 
            logger.error('RBAC requires a user');
            return false; 
        };
```

Reply:
I have decided a bit simpler approach. First check that the user exists. Then it is safe to assume it exist later on.